6484 Update plugin documentation (#6485)
msupply-foundation/open-msupply: client/packages/plugins/README.md

In production mode the process differs:

- the server provides an endpoint to fetch the list of available plugins
- the client app fetches the full list on startup, then individually fetches each plugin bundle
- Using federation module and webpack low level api the plugin is dynamically added to scope
- When plugin and it's dependencies are resolved it's added to PluginProvider and it becomes available to be used by the core functionality

### Plugin definitions

> Plugin version and plugin code are defined in package.json of the plugin, code should be unique across all plugins, so it's a good idea to have something unique in the code. 

Reviewer: Per suggestion in other PR, I think we would benefit distinguishing `plugin` and `plugin bundle`.

The submodules that we load into `plugin` folder should always be referred to as `plugin bundles`, they define sets of plugins that are merged by the plugin provider

So this section would be `Plugin Bundle definitions` etc...

Reviewer: Or, if `bundle` is already an overloaded term given the use of webpack, and the frontend/backend bundling, i could also go for `plugin set`? But I do think we need a distinction, especially since this is intended as external developer facing!

Author: Hmm, i feel like plugin is a combination of functionality change, not a single interface override. I hear you, I think if we are changing in docs we should change in most places, I would like to make that change but not sure how and where (do we rename everything plugins bundle, like the 'plugins' export, should that be pluginBundle ?).
Keen to chat about this

`usePluginEvents` helper can store any state and provide any type to event listener which in turn can return any type, for example if you want to set `isReady` state and trigger an event `onSave` that passes in a `string` to be validated, expecting `error` if not valid then:

```typescript
// ### in Core 
// When defining plugin interface
events: UsePluginEvents<{ isReady: boolean }, { validateThisString: string }, 'ok' | { error: 'string' }>;
//  Bind to component
const pluginEvents = usePluginEvents<_, { validateThisString: string }, 'ok' | { error: 'string' }>({ isReady: false });
// When asking for validation
const validate = async(validateThisString: string) => {
   let validationResult = await pluginEvents.dispatchEvent({ validateThisString: 'good'});
   if (validationResult == 'ok') {
      closeModal()
   }
   error(validationResult.error) // This is a toast
   setError(validationResult.error)
}
// Checking for isReady
<div>
  {pluginEvents.state.isReady && {'i am ready'}}

> </div>

Author: Thanks, i merged it in

        }
        return { error: 'string is not good' }
    }
  );
  return unmountEvent; // mountEvent return a handler to unmountEvent
}, []);
// Setting isReady 

useEffect(() => {
  events.setState({isReady: true})
}, [somethingChanged])
```

Of course in above example the types should be defined once and shared (even though typescript guarantees they are valid it's a bit verbose)

TODO actually tried to share types but had typescript error when doing `const pluginEvents: SomeConcreteTypeOfUsePluginEvent = usePluginEvents>({ isReady: false });` <- said that (boolean is not false ¯\_(ツ)_/¯)

### Plugin data

TODO update this

Plugins can store data in the `plugin_data` table. The following methods are available in the graphQL API for interacting with plugin data:

- `pluginData`
- `insertPluginData`
- `updatePluginData`

The querying and mutating of data follows the standard pattern used throughout open mSupply:

```typescript
  const { data } = usePluginData.data(stockLine?.id ?? '');
  const { mutate } = data?.id ? usePluginData.update() : usePluginData.insert();
```

These functions can be implemented within your plugin and used to fetch and update data.

> ## Creating a plugin

Reviewer: Would be good to mention in this section that hot module reloading is available for local dev, but frontend will need to be restarted once on initial addition of the plugin bundle submodule/folder

Author: a2aee2c0fbe71a96883c9aa465589e657c0eef65

You can watch [this video for example](https://drive.google.com/file/d/1JnmPU9hRaQD4R1hTDKbbNj78FnM2l00A/view?usp=drive_link) TODO make public

> The simplest way to begin is by cloning (forking for now or just copy and create new repo, until we have a template), this repository https://github.com/msupply-foundation/open-msupply-plugin, then add it as submodule to `client/packages/plugins/` using this command: `git submodule add https://github.com/andreievg/open-msupply-plugins-andrei client/packages/plugins/mynewplugin`, note the `mynewplugin` can be anything. The inner repository and core will be treated as two different repositories, changes in them will only be reflected in relative repositories (i.e. you can add the inner repository as local repository in github desktop). Make sure that you don't commit the `.gitmodule` or the single `client/packages/plugins/{your plugin name}` to the core.

Reviewer: ```suggestion
The simplest way to begin is by cloning (forking for now or just copy and create new repo, until we have a template), this repository https://github.com/msupply-foundation/open-msupply-plugins, then add it as a submodule to `client/packages/plugins/`. From the root of this repository, run: 

```git submodule add [your-plugin-bundle-repo-url] client/packages/plugins/myPluginBundle```

Note the `myPluginBundle` can be anything. The inner repository and core will be treated as two different repositories, changes in them will only be reflected in relative repositories (i.e. you can add the inner repository as local repository in github desktop). Make sure that you don't commit the `.gitmodule` or the single `client/packages/plugins/{your plugin bundle name}` to the core.
```

Author: Thanks
